pypuf/learner/evolution_strategies/becker.py

```diff
@@ -0,0 +1,311 @@
+        self.path_cm = np.matrix(np.zeros(instance.n)) # cumulated evolution path of covariance matrix
+        self.path_ss = np.zeros(instance.n) # cumulated evolution path of step size
+        # auxiliary constants
+        self.mu_w = 1 / np.sum(np.square(priorities))
+        self.c_mu = self.mu_w / instance.n**2
+        self.d_sigma = 1 + np.sqrt(self.mu_w / instance.n)
+        self.c_1 = 2 / instance.n**2
+        self.c_c = 4 / instance.n
+        self.c_sigma = 4 / instance.n
+        assert len(priorities) == parent_size
+        assert int(np.ndarray.sum(priorities)) == 1
+        assert (self.c_1 + self.c_mu <= 1)
+
+    def learn(self):
+        # this is the main learning method
+        # returns XOR-LTFArray with nearly same behavior as learned instance
+        self.epsilon = self.learn_epsilon(self.instance, self.unreliability)
+        while self.num_of_LTFs < self.instance.k:
+            self.new_LTF = self.learn_LTF()
+            if self.is_different_LTF():
+                self.different_LTFs[self.num_of_LTFs] = self.new_LTF
+                self.num_of_LTFs += 1
+        self.different_LTFs = self.set_pole_of_LTFs(self.instance, self.different_LTFs)
+        return LTFArray(self.different_LTFs, self.transform, self.combiner, bias=False)
+
+    def is_different_LTF(self):
+        # returns True iff new_LTF is different from previously learned LTFs
+        if self.num_of_LTFs == 0 :
+            return True
+        weight_arrays = self.different_LTFs[:self.num_of_LTFs-1, :]
+        new_LTFArray = LTFArray(self.new_LTF, self.transform, self.combiner)
+        different_LTFs = self.build_LTFArrays(weight_arrays)
+        challenges = tools.sample_inputs(self.instance.n, self.challenge_num)
+        responses = np.empty((self.num_of_LTFs, self.challenge_num))
+        for i, challenge in enumerate(challenges):
+            responses[0, i] = new_LTFArray.eval(challenge)
+            for j, current_LTF in enumerate(different_LTFs):
+                responses[j+1, i] = current_LTF.eval(challenge)
+        return not self.is_correlated(responses)
+
+    def learn_LTF(self):
+        # this is the main CMA-ES algorithm like that from Hansen
+        terminate = False
+        while not terminate:
+            challenges = tools.sample_inputs(
+                    self.instance.n, self.challenge_num)
+            measured_rels = np.zeros(self.challenge_num)
+            while np.sum(measured_rels) < (1-self.unreliability)*self.challenge_num:
+                challenges = tools.sample_inputs(
+                    self.instance.n, self.challenge_num)
+                measured_rels = self.measure_rels(
+                    self.instance, challenges, self.challenge_num, self.repeat,
+                    self.epsilon)
+            sorted_individuals = self.fitness(
+                challenges, self.challenge_num, measured_rels,
+                self.individuals, self.epsilon)
+            if self.terminate:
```

Comment:
There is two termination variables, `terminate` and `self.terminate`, which is kind of confusing. Refactor `self.terminate` into a function return value, if possible